Algorithm.Framework/Portfolio/MinimumVariancePortfolioOptimizer.cs

```diff
@@ -0,0 +1,100 @@
+using Accord.Math;
+using Accord.Math.Optimization;
+using Accord.Statistics;
+
+namespace QuantConnect.Algorithm.Framework.Portfolio
+{
+    /// <summary>
+    /// Provides an implementation of a minimum variance portfolio optimizer that calculate the optimal weights 
+    /// with the weight range from -1 to 1 and minimize the portfolio variance with a target return of 2%
+    /// </summary>
+    public class MinimumVariancePortfolioOptimizer : IPortfolioOptimizer
+    {
+        protected double _lower;
+        protected double _upper;
+        protected double _targetReturn;
+        protected double[,] _cov;
+        protected List<LinearConstraint> _constraints;
+
+        protected int Size => _cov == null ? 0 : _cov.GetLength(0);
+
+        public MinimumVariancePortfolioOptimizer(double lower = -1, double upper = 1, double targetReturn = 0.02)
+        {
+            _lower = lower;
+            _upper = upper;
+            _targetReturn = targetReturn;
+            _cov = null;
+            _constraints = new List<LinearConstraint>();
+        }
+
+        protected double[] Optimize()
+        {
+            // sum(x) = 1
+            _constraints.Add(new LinearConstraint(Size)
+            {
+                CombinedAs = Vector.Create(Size, 1.0),
+                ShouldBe = ConstraintType.EqualTo,
+                Value = 1.0
+            });
+
+            // lw <= x <= up
+            for (int i = 0; i < Size; i++)
+            {
+                _constraints.Add(new LinearConstraint(1)
+                {
+                    VariablesAtIndices = new int[] { i },
+                    ShouldBe = ConstraintType.GreaterThanOrEqualTo,
+                    Value = _lower
+                });
+                _constraints.Add(new LinearConstraint(1)
+                {
+                    VariablesAtIndices = new int[] { i },
+                    ShouldBe = ConstraintType.LesserThanOrEqualTo,
+                    Value = _upper
+                });
+            }
+
+            var optfunc = new QuadraticObjectiveFunction(_cov, Vector.Create(Size, 0.0));
+            var solver = new GoldfarbIdnani(optfunc, _constraints);
+            _constraints.Clear();
+
+            var init = Vector.Create(Size, 1.0 / Size);
+            bool success = solver.Minimize(init);
+            return success ? solver.Solution : init;
+        }
+
+        public virtual double[] Optimize(double[,] historicalReturns, double[] expectedReturns = null)
+        {
+            _cov = historicalReturns.Covariance();
+            var returns = expectedReturns ?? historicalReturns.Mean(0);
+
+            // mu^T x = R  or mu^T x >= 0
+            _constraints.Add(new LinearConstraint(Size)
+            {
+                CombinedAs = returns,
+                ShouldBe = _targetReturn > 0 ? ConstraintType.EqualTo : ConstraintType.GreaterThanOrEqualTo,
+                Value = _targetReturn
+            });
+
+            return Optimize();
```

Comment:
`IPortfolioOptimizer.Optimize` should not be used as a method to set all the private class fields. Why are `_cov` and `_constraints` private class fields?
```csharp
public virtual double[] Optimize(double[,] historicalReturns, double[] expectedReturns = null)
{
    var cov = historicalReturns.Covariance();
    var returns = expectedReturns ?? historicalReturns.Mean(0);
    var constraints = new List<LinearConstraint>();
    var size = cov..GetLength(0);  

    constraints.Add(GetBudgetConstraint());
    constraints.AddRange(GetBoundaryConditions(size));
    // etc
    var optfunc = new QuadraticObjectiveFunction(cov, Vector.Create(size, 0.0));
    var solver = new GoldfarbIdnani(optfunc, constraints);
    constraints.Clear();

    var init = Vector.Create(Size, 1.0 / Size);
    var success = solver.Minimize(init);
    return success ? solver.Solution : init;
}

Reply:
> Why are `_cov` and `_constraints` private class fields?

Actually, we solve MVO QP which is something like this:
```
public virtual double[] Optimize(double[,] historicalReturns, double[] expectedReturns = null)
{   
    var cov = historicalReturns.Covariance();
    var returns = expectedReturns ?? historicalReturns.Mean(0);
    var constraints = new List<LinearConstraint>();
    var size = cov.GetLength(0);  

   // some additional constraints
   // i.e. sum(x) = 1

    return Optimize(cov, constraints);
}
```
This part of problem constraction is the same for both MVO and MSR algorithms:
```
public virtual double[] Optimize(cov, constraints)
    var size = cov.GetLength(0);  
    constraints.Add(GetBudgetConstraint());
    constraints.AddRange(GetBoundaryConditions(size));
    var optfunc = new QuadraticObjectiveFunction(cov, Vector.Create(size, 0.0));
    var solver = new GoldfarbIdnani(optfunc, constraints);
    constraints.Clear();

    var init = Vector.Create(Size, 1.0 / Size);
    var success = solver.Minimize(init);
    return success ? solver.Solution : init;
}
```

Reply:
@wildart, could you please keep the models independent like we did in python? I mean, let's not inherit one from another. These model are meant for educational purposes and it doubles the teaching material the way you have done it which is great from the code maintenance point of view.

Reply:
No problem.